crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -2770,9 +2770,17 @@ impl<'a> Checker<'a> {
 
                         self.semantic.restore(snapshot);
 
-                        if self.semantic.in_typing_only_annotation() {
+                        let add_future_import = self.settings().allow_importing_future_annotations
+                            && self.semantic.in_runtime_evaluated_annotation();
+
+                        if self.semantic.in_typing_only_annotation() || add_future_import {
                             if self.is_rule_enabled(Rule::QuotedAnnotation) {
-                                pyupgrade::rules::quoted_annotation(self, annotation, range);
+                                pyupgrade::rules::quoted_annotation(
+                                    self,
+                                    annotation,
+                                    range,
+                                    add_future_import,
+                                );
                             }
                         }
```

Comment:
Nit: I would probably move the entire condition into `quoted_annotation`, considering that `quoted_annotation` already takes `self` as an argument (it has everything it needs to determine itself if it should put this into quotes or not). 
```suggestion
                        if self.is_rule_enabled(Rule::QuotedAnnotation) {
                            pyupgrade::rules::quoted_annotation(
                                self,
                                annotation,
                                range,
                            );
                        }
```